### ISSUE TITLE
Subhandler states are not cleared at the end of handling cycle

### ISSUE THREAD

**Issue by MarkusH** [zalando-incubator/kopf#384]
## Long story short

When using `kopf.register()` or `kopf.execute()`, the state of the "root" handler is tracked in `kopf.reactor.processing.process_resource_changing_cause()` which calls [`state.store()`](https://github.com/zalando-incubator/kopf/blob/349a976f65599832ac91db6bc91d14365bf497ba/kopf/reactor/processing.py#L381). It also calls [`state.purge()`](https://github.com/zalando-incubator/kopf/blob/349a976f65599832ac91db6bc91d14365bf497ba/kopf/reactor/processing.py#L386) a few lines below, when `state.done`. For sub-handlers, there is a [`state.store()`](https://github.com/zalando-incubator/kopf/blob/349a976f65599832ac91db6bc91d14365bf497ba/kopf/reactor/handling.py#L170) call in `kopf.reactor.handling.execute()`. However, the state for sub-handlers is never purged.

## Description

The failing purging of sub-handler states leads to sub-handlers not being executed. I'm not sure if that has implications or may even be the cause for #279.

```python
import functools

import kopf


async def subhandler1(**kwargs):
    pass


async def subhandler2(**kwargs):
    pass


@kopf.on.update("example.com", "v1", "as")
async def update(**kwargs):
    await kopf.execute(
        fns={"s1": functools.partial(subhandler1), "s2": functools.partial(subhandler2)}
    )
```

```yaml
apiVersion: example.com/v1
kind: A
metadata:
  annotations:
    kopf.zalando.org/last-handled-configuration: '{"spec": {"a": 1}}'
  creationTimestamp: "2020-07-11T16:19:28Z"
  generation: 1
  name: test
  namespace: default
  resourceVersion: "4657318"
  selfLink: /apis/example.com/v1/namespaces/default/as/test
  uid: a79a6ca3-687c-4467-abc1-a38f0019a5ca
spec:
  a: 1
```

After triggering a change:

```yaml
apiVersion: example.com/v1
kind: A
metadata:
  annotations:
    kopf.zalando.org/last-handled-configuration: '{"spec": {"a": 2}}'
    kopf.zalando.org/update.s1: '{"started": "2020-07-11T16:20:02.463066", "stopped":
      "2020-07-11T16:20:02.464577", "retries": 1, "success": true, "failure": false}'
    kopf.zalando.org/update.s2: '{"started": "2020-07-11T16:20:02.463137", "stopped":
      "2020-07-11T16:20:02.737845", "retries": 1, "success": true, "failure": false}'
  creationTimestamp: "2020-07-11T16:19:28Z"
  generation: 4
  name: test
  namespace: default
  resourceVersion: "4657369"
  selfLink: /apis/example.com/v1/namespaces/default/as/test
  uid: a79a6ca3-687c-4467-abc1-a38f0019a5ca
spec:
  a: 2
status:
  kopf:
    progress:
      update/s1:
        failure: false
        retries: 1
        started: 2020-07-11T16:20:02.463066
        stopped: 2020-07-11T16:20:02.464577
        success: true
      update/s2:
        failure: false
        retries: 1
        started: 2020-07-11T16:20:02.463137
        stopped: 2020-07-11T16:20:02.737845
        success: true
```

```
[2020-07-11 18:20:02,460] kopf.objects         [DEBUG   ] [default/test] Update event: {'apiVersion': 'example.com/v1', 'kind': 'A', 'metadata': {'annotations': {'kopf.zalando.org/last-handled-configuration': '{"spec": {"a": 1}}'}, 'creationTimestamp': '2020-07-11T16:19:28Z', 'generation': 2, 'name': 'test', 'namespace': 'default', 'resourceVersion': '4657365', 'selfLink': '/apis/example.com/v1/namespaces/default/as/test', 'uid': 'a79a6ca3-687c-4467-abc1-a38f0019a5ca'}, 'spec': {'a': 2}}
[2020-07-11 18:20:02,460] kopf.objects         [DEBUG   ] [default/test] Update diff: (('change', ('spec', 'a'), 1, 2),)
[2020-07-11 18:20:02,462] kopf.objects         [DEBUG   ] [default/test] Handler 'update' is invoked.
[2020-07-11 18:20:02,463] kopf.objects         [DEBUG   ] [default/test] Handler 'update/s1' is invoked.
[2020-07-11 18:20:02,464] kopf.objects         [INFO    ] [default/test] Handler 'update/s1' succeeded.
[2020-07-11 18:20:02,465] kopf.objects         [DEBUG   ] [default/test] Handler 'update' has unfinished sub-handlers. Will retry soon.
[2020-07-11 18:20:02,465] kopf.objects         [DEBUG   ] [default/test] Patching with: {'metadata': {'annotations': {'kopf.zalando.org/update.s1': '{"started": "2020-07-11T16:20:02.463066", "stopped": "2020-07-11T16:20:02.464577", "retries": 1, "success": true, "failure": false}', 'kopf.zalando.org/update.s2': '{"started": "2020-07-11T16:20:02.463137", "retries": 0, "success": false, "failure": false}', 'kopf.zalando.org/update': '{"started": "2020-07-11T16:20:02.461880", "delayed": "2020-07-11T16:20:02.465470", "retries": 1, "success": false, "failure": false, "message": "None"}'}}, 'status': {'kopf': {'progress': {'update/s1': {'started': '2020-07-11T16:20:02.463066', 'stopped': '2020-07-11T16:20:02.464577', 'delayed': None, 'retries': 1, 'success': True, 'failure': False, 'message': None}, 'update/s2': {'started': '2020-07-11T16:20:02.463137', 'stopped': None, 'delayed': None, 'retries': 0, 'success': False, 'failure': False, 'message': None}, 'update': {'started': '2020-07-11T16:20:02.461880', 'stopped': None, 'delayed': '2020-07-11T16:20:02.465470', 'retries': 1, 'success': False, 'failure': False, 'message': 'None'}}}}}
[2020-07-11 18:20:02,733] kopf.objects         [DEBUG   ] [default/test] Update event: {'apiVersion': 'example.com/v1', 'kind': 'A', 'metadata': {'annotations': {'kopf.zalando.org/last-handled-configuration': '{"spec": {"a": 1}}', 'kopf.zalando.org/update': '{"started": "2020-07-11T16:20:02.461880", "delayed": "2020-07-11T16:20:02.465470", "retries": 1, "success": false, "failure": false, "message": "None"}', 'kopf.zalando.org/update.s1': '{"started": "2020-07-11T16:20:02.463066", "stopped": "2020-07-11T16:20:02.464577", "retries": 1, "success": true, "failure": false}', 'kopf.zalando.org/update.s2': '{"started": "2020-07-11T16:20:02.463137", "retries": 0, "success": false, "failure": false}'}, 'creationTimestamp': '2020-07-11T16:19:28Z', 'generation': 3, 'name': 'test', 'namespace': 'default', 'resourceVersion': '4657366', 'selfLink': '/apis/example.com/v1/namespaces/default/as/test', 'uid': 'a79a6ca3-687c-4467-abc1-a38f0019a5ca'}, 'spec': {'a': 2}}, 'status': {'kopf': {'progress': {'update': {'delayed': '2020-07-11T16:20:02.465470', 'failure': False, 'message': 'None', 'retries': 1, 'started': '2020-07-11T16:20:02.461880', 'success': False}, 'update/s1': {'failure': False, 'retries': 1, 'started': '2020-07-11T16:20:02.463066', 'stopped': '2020-07-11T16:20:02.464577', 'success': True}, 'update/s2': {'failure': False, 'retries': 0, 'started': '2020-07-11T16:20:02.463137', 'success': False}}}}}
[2020-07-11 18:20:02,734] kopf.objects         [DEBUG   ] [default/test] Update diff: (('change', ('spec', 'a'), 1, 2),)
[2020-07-11 18:20:02,735] kopf.objects         [DEBUG   ] [default/test] Handler 'update' is invoked.
[2020-07-11 18:20:02,736] kopf.objects         [DEBUG   ] [default/test] Handler 'update/s2' is invoked.
[2020-07-11 18:20:02,737] kopf.objects         [INFO    ] [default/test] Handler 'update/s2' succeeded.
[2020-07-11 18:20:02,738] kopf.objects         [INFO    ] [default/test] Handler 'update' succeeded.
[2020-07-11 18:20:02,739] kopf.objects         [INFO    ] [default/test] All handlers succeeded for update.
[2020-07-11 18:20:02,740] kopf.objects         [DEBUG   ] [default/test] Patching with: {'metadata': {'annotations': {'kopf.zalando.org/update.s2': '{"started": "2020-07-11T16:20:02.463137", "stopped": "2020-07-11T16:20:02.737845", "retries": 1, "success": true, "failure": false}', 'kopf.zalando.org/update': None}}, 'status': {'kopf': {'progress': {'update/s2': {'started': '2020-07-11T16:20:02.463137', 'stopped': '2020-07-11T16:20:02.737845', 'delayed': None, 'retries': 1, 'success': True, 'failure': False, 'message': None}, 'update': None}}}}
[2020-07-11 18:20:02,883] kopf.objects         [DEBUG   ] [default/test] Something has changed, but we are not interested (the essence is the same).
[2020-07-11 18:20:02,883] kopf.objects         [DEBUG   ] [default/test] Handling cycle is finished, waiting for new changes since now.
```

Making yet another update to the custom resource:

```yaml
apiVersion: example.com/v1
kind: A
metadata:
  annotations:
    kopf.zalando.org/last-handled-configuration: '{"spec": {"a": 3}}'
    kopf.zalando.org/update.s1: '{"started": "2020-07-11T16:20:02.463066", "stopped":
      "2020-07-11T16:20:02.464577", "retries": 1, "success": true, "failure": false}'
    kopf.zalando.org/update.s2: '{"started": "2020-07-11T16:20:02.463137", "stopped":
      "2020-07-11T16:20:02.737845", "retries": 1, "success": true, "failure": false}'
  creationTimestamp: "2020-07-11T16:19:28Z"
  generation: 5
  name: test
  namespace: default
  resourceVersion: "4748091"
  selfLink: /apis/example.com/v1/namespaces/default/as/test
  uid: a79a6ca3-687c-4467-abc1-a38f0019a5ca
spec:
  a: 3
status:
  kopf:
    progress:
      update/s1:
        failure: false
        retries: 1
        started: 2020-07-11T16:20:02.463066
        stopped: 2020-07-11T16:20:02.464577
        success: true
      update/s2:
        failure: false
        retries: 1
        started: 2020-07-11T16:20:02.463137
        stopped: 2020-07-11T16:20:02.737845
        success: true
```

```
[2020-07-12 11:46:28,666] kopf.objects         [DEBUG   ] [default/test] Update event: {'apiVersion': 'example.com/v1', 'kind': 'A', 'metadata': {'annotations': {'kopf.zalando.org/last-handled-configuration': '{"spec": {"a": 2}}', 'kopf.zalando.org/update.s1': '{"started": "2020-07-11T16:20:02.463066", "stopped": "2020-07-11T16:20:02.464577", "retries": 1, "success": true, "failure": false}', 'kopf.zalando.org/update.s2': '{"started": "2020-07-11T16:20:02.463137", "stopped": "2020-07-11T16:20:02.737845", "retries": 1, "success": true, "failure": false}'}, 'creationTimestamp': '2020-07-11T16:19:28Z', 'generation': 5, 'name': 'test', 'namespace': 'default', 'resourceVersion': '4748090', 'selfLink': '/apis/example.com/v1/namespaces/default/as/test', 'uid': 'a79a6ca3-687c-4467-abc1-a38f0019a5ca'}, 'spec': {'a': 3}}, 'status': {'kopf': {'progress': {'update/s1': {'failure': False, 'retries': 1, 'started': '2020-07-11T16:20:02.463066', 'stopped': '2020-07-11T16:20:02.464577', 'success': True}, 'update/s2': {'failure': False, 'retries': 1, 'started': '2020-07-11T16:20:02.463137', 'stopped': '2020-07-11T16:20:02.737845', 'success': True}}}}}
[2020-07-12 11:46:28,666] kopf.objects         [DEBUG   ] [default/test] Update diff: (('change', ('spec', 'a'), 2, 3),)
[2020-07-12 11:46:28,668] kopf.objects         [DEBUG   ] [default/test] Handler 'update' is invoked.
[2020-07-12 11:46:28,669] kopf.objects         [INFO    ] [default/test] Handler 'update' succeeded.
[2020-07-12 11:46:28,670] kopf.objects         [INFO    ] [default/test] All handlers succeeded for update.
[2020-07-12 11:46:28,671] kopf.objects         [DEBUG   ] [default/test] Patching with: {'metadata': {'annotations': {'kopf.zalando.org/last-handled-configuration': '{"spec": {"cluster": {"imageRegistry": "crate", "license": {"secretKeyRef": {"key": "license-key", "name": "cratedb-license"}}, "name": "my-crate-cluster", "version": "4.1.5"}, "nodes": {"data": [{"name": "t1", "replicas": 8, "resources": {"cpus": 0.25, "disk": {"count": 1, "size": "8GiB", "storageClass": "default"}, "heapRatio": 0.5, "memory": "512Mi"}}]}}}'}}}
[2020-07-12 11:46:28,921] kopf.objects         [DEBUG   ] [default/test] Something has changed, but we are not interested (the essence is the same).
[2020-07-12 11:46:28,921] kopf.objects         [DEBUG   ] [default/test] Handling cycle is finished, waiting for new changes since now.
```

## Environment

* Kopf version: 0.27

**Comment by nemethf** [comment]
I think I ran into this as well.  My workaround was to append kwarg["started"] to the id of the kopf.on.this decorator:
```python
        @kopf.on.this(id=f's1-kwargs["started"]')
        async def subhandler1(**kwargs):
            pass
```
With this workaround, the sub-hander runs on each update, but kopf now litters the annotations of the custom resource. I hope PR #384 fixes this issue.  Thanks, @MarkusH, for that.

**Comment by MarkusH** [comment]
Hi @nolar, I can only imagine how busy you must be. Would you be able to look into the PR (#385) and let me know if the approach taken there makes sense? I'm happy to invest more time to get this problem fixed. Thank you. :slightly_smiling_face:

**Comment by nolar** [comment]
Solved in https://github.com/nolar/kopf/pull/517, available since `kopf>=0.28rc2`
